p2p/client.go

```diff
@@ -77,15 +79,23 @@ type Client struct {
 //
 // Basic checks on parameters are done, and default parameters are provided for unset-configuration
 // TODO(tzdybal): consider passing entire config, not just P2P config, to reduce number of arguments
-func NewClient(conf config.P2PConfig, privKey crypto.PrivKey, chainID string, logger log.Logger) (*Client, error) {
+func NewClient(conf config.P2PConfig, privKey crypto.PrivKey, chainID string, gater *conngater.BasicConnectionGater, logger log.Logger) (*Client, error) {
 	if privKey == nil {
 		return nil, errNoPrivKey
 	}
 	if conf.ListenAddress == "" {
 		conf.ListenAddress = config.DefaultListenAddress
 	}
+	if gater == nil {
```

Comment:
I'm not a big fan of this. IMHO, if `nil` is passed, we should just not use gater at all.

Reply:
Hmm, then let's make having a ConnectionGater mandatory, because making it optional makes `Client.listen()` ugly.

It won't do anything differently if there are no rulesets.

Reply:
This does, however, make anything that uses `NewClient()` more ugly as it has to prepare the `gater` for `NewClient()`.

Reply:
I kept ConnectionGater optional.

```
func (c *Client) listen(ctx context.Context) (host.Host, error) {
	var err error
	maddr, err := multiaddr.NewMultiaddr(c.conf.ListenAddress)
	if err != nil {
		return nil, err
	}

	var host host.Host
	if c.gater == nil {
		host, err = libp2p.New(libp2p.ListenAddrs(maddr), libp2p.Identity(c.privKey))
	} else {
		host, err = libp2p.New(libp2p.ListenAddrs(maddr), libp2p.Identity(c.privKey), libp2p.ConnectionGater(c.gater))
	}
	if err != nil {
		return nil, err
	}

	return host, nil
}
```